carla_ros_bridge/src/carla_ros_bridge/ego_traffic_light_sensor.py

```diff
@@ -0,0 +1,178 @@
+        self.stop_line_info_map = None
+        self.node.loginfo("Destroy EgoTrafficLightSensor")
+
+    @staticmethod
+    def get_blueprint_name():
+        return "sensor.pseudo.ego_traffic_light"
+
+    def update(self, frame, timestamp):
+        try:
+            if self.ego is None or self.map is None:
+                return
+
+            if self.stop_line_info_map is None or self.map_name != self.map.name:
+                self.stop_line_info_map = create_stop_line_info_map(self.node, self.actor_list)
+                self.map_name = self.map.name
+
+            ego_location = self.ego.get_location()
+            wp = self.map.get_waypoint(ego_location)
+            sli = get_stop_line_info(self.stop_line_info_map, wp.road_id, wp.lane_id)
+
+            inside_intersection = wp.is_junction
+
+            if (self.msg.inside_intersection != inside_intersection
+                or self.has_traffic_light_status_changes(self.cur_sli, sli)
+                or self._info_published_at is None
+                or timestamp - self._info_published_at > _PUBLISH_INTERVAL_SECONDS
+            ):
+                self.calculate_and_publish_data(ego_location, sli, inside_intersection)
+                self._info_published_at = timestamp
+        except Exception as e:
+            self.node.loginfo("Error: {}".format(e))
+
+    def calculate_and_publish_data(self, ego_location: carla.Location, new_sli: StoplineInfo, inside_intersection: bool):
+        """Publish CarlaEgoTrafficLightInfo message."""
+        if inside_intersection:
+            self.msg.distance_to_stopline = -1.0
+            if self.cur_sli:
+                self.msg.traffic_light_status = self.cur_sli.traffic_light_actor.get_status()
+        else:
+            if new_sli:
+                self.msg.distance_to_stopline = ego_location.distance(new_sli.stop_location)
+                self.msg.traffic_light_status = new_sli.traffic_light_actor.get_status()
+            else:
+                self.msg.distance_to_stopline = -1.0
+                self.msg.traffic_light_status = CarlaTrafficLightStatus()
+            # Store new StopLineInfo.
+            self.cur_sli = new_sli
+
+        self.msg.inside_intersection = inside_intersection
+        self.pub.publish(self.msg)
+
+    def get_ego(self, actor_list: Dict[int, Actor]):
+        """Return a CarlaActor representing the ego car."""
+        for actor in actor_list.values():
+            if isinstance(actor, EgoVehicle):
+                return actor.carla_actor
+        return None
+
+    def has_traffic_light_status_changes(self, cur_sli: StoplineInfo, new_sli: StoplineInfo):
+        """Return True if the traffic light status (either new traffic light id or new state) has been changed; False otherwise."""
+        if cur_sli is None or new_sli is None:
+            return False
+
+        cur_status = cur_sli.traffic_light_actor.get_status()
+        new_status = new_sli.traffic_light_actor.get_status()
```

Comment:
would it not be enough to publish the lanes aren't the same? why don't we want to publish if the lights are the same but the lines changed? or am i missing that two traffic lights of the same color will fail the sameness check

Reply:
When the ego is inside the intersection, the StopLineInfo is not updated yet.

I passed in the traffic light status instead of the StopLineInfo. I think that will be clearer.